### PR TITLE
Adding last week to july 19

### PR DESCRIPTION
**Pull request by chylauSF**
## Problem

As we are adding a 5th week in January FY19 and adding more dates to the `core_fiscal_year`, we are expecting the MerchCalendar gem to reflect that change, and to hopefully use it to regenerate the dates for future fiscal years.

## Solution
I changed the specs to reflect the correct output of what an FY19th year should look like.  Example:  the beginning of FY 19 should start, `7/29/18` and end in `8/3/2019`.  The previous code, assumed that FY18 was a 53rd week year, so I had to change the its assumptions and made sure we had the correct calculations for FY18 to have 52 weeks and FY19 to have 53 weeks.

What I discovered is that the ending date of any calendar (Fiscal or Merch) was critical in how to count how many weeks within a year.  Determining the [year_end](https://github.com/stitchfix/merch_calendar/blob/master/lib/merch_calendar/retail_calendar.rb#L11) in the `end of year` method was so critical in calculating the correct number of weeks, that the only significant change to differentiate a Merch Calendar with a Fiscal one is changing the `year_end` from January 31st (year end for retail/merch calendar) to July 31st (year end for fiscal).  

## Deployment checklist
- [ ] Use this gem to generate more dates for Fiscal Year 2019 and so on...
- [ ] Design a Rake task to do above.
### Pre-deployment
- [ ] I desire to refactor this code... because I recognize that we are essentially using the same logic, just different `year_end`s 
- [ ] Awaiting Jon's review, and advice on how to refactor
- [ ] Maybe add specs for future years passing FY19 to make sure we're consistent for future dates

## Reviewers:
@jonworek,  I definitely need your advice on how refactor the code.  Let me know if a call is better to explain how I should reduce the redundancy of the code.

### CODE REVIEW

File: lib/merch_calendar/fiscal_year_calendar.rb [hunk 1]
```diff
@@ -86,18 +86,5 @@ def weeks_in_year(year)
       ((start_of_year(year + 1) - start_of_year(year)) / 7).to_i
     end
     
-    def merch_months_in(start_date, end_date)
```
Comment (jonworek) [inline lib/merch_calendar/fiscal_year_calendar.rb]: Why did you remove this method?

File: lib/merch_calendar/retail_calendar.rb [hunk 2]
```diff
@@ -27,7 +27,7 @@ def start_of_year(year)
     # The starting date of a given month
     # THIS IS THE MERCH MONTH
     # 1 = feb
-    # 
+
```
Comment (jonworek) [inline lib/merch_calendar/retail_calendar.rb]: Comments about a method should be right on top of the method, no blank lines in between.

File: lib/merch_calendar/retail_calendar.rb [hunk 3]
```diff
@@ -8,12 +8,12 @@ class RetailCalendar
     QUARTER_4 = 4
 
     def end_of_year(year)
-      year_end = Date.new((year + 1), 1, -1)
-      wday = (year_end.wday + 1) % 7
+      year_end = Date.new((year + 1), 1, -1) # Jan 31st
+      wday = (year_end.wday + 1) % 7 
 
-      if wday > 3
+      if wday > 3 ### this rounds up to the next saturday
```
Comment (jonworek) [inline lib/merch_calendar/retail_calendar.rb]: You have two more '#' signs than you need.

File: lib/merch_calendar/fiscal_year_calendar.rb [hunk 4]
```diff
@@ -2,117 +2,89 @@
...
+        end_of_month(year, 6)
+      when QUARTER_3
+        end_of_month(year, 9)
+      when QUARTER_4
+        end_of_month(year, 12)
+      end
     end
 
-    private
+    def end_of_year(year)
```
Comment (jonworek) [inline lib/merch_calendar/fiscal_year_calendar.rb]: Reordering methods, and changing their behavior, makes it really hard on a reviewer to see what changed.  Can you move them back to their original ordering?

If you want to rearrange, I'd suggest doing it in another PR.

File: lib/merch_calendar/fiscal_year_calendar.rb [hunk 5]
```diff
@@ -2,117 +2,89 @@
...
-        [year, month + month_offset] 
-      else
-        [year - 1, month - month_offset]
+    
+    def start_of_month(year, merch_month)
+      # 91 = number of days in a single 4-5-4 set 
+      start = start_of_year(year) + ((merch_month - 1) / 3).to_i * 91
+
+      case merch_month
+      when 2,5,8,11
```
Comment (jonworek) [inline lib/merch_calendar/fiscal_year_calendar.rb]: The code in the retail calendar should not be taken as a good example, as it uses a lot of "magic numbers".  Since you're rewriting this, anywhere you see a number, please extract it to a constant.

**Comment by jonworek** [comment]
From your PR description:

>Problem
>
>stitchfix/erch#1848
>stitchfix/erch#1892
>
>As we are adding a 5th week in January FY19 and adding more dates to the core_fiscal_year, we are expecting the MerchCalendar gem to reflect that change, and to hopefully use it to regenerate the dates for future fiscal years.
>
>Deployment checklist
>
> Use this gem to generate more dates for Fiscal Year 2019 and so on...
> Design a Rake task to do above.

Things like this shouldn't be mentioned in this repository.  These are internal details of work we need to do at Stitch Fix, and this is a public, open source library.

PRs in this repo should be written in a manner that anyone outside of Stitch Fix can understand.

Not to say that you're risking leaking confidential info, but keep in mind your audience.  The things you mention above will be handled in our internal apps, i.e. fashionthing.

**Comment by jonworek** [comment]
One last thing... you may want to take a look into why your code coverage score went down:

screen shot 2017-08-15 at 9 08 07 am

If you're not familiar, this indicates that your tests covered more code before your changes, than after your changes.

File: lib/merch_calendar/fiscal_year_calendar.rb (hunk 1 above)
Comment (chylauSF) [inline lib/merch_calendar/fiscal_year_calendar.rb]: I wasn't sure what it was doing?  But I can add it back in.

File: lib/merch_calendar/fiscal_year_calendar.rb (hunk 4 above)
Comment (chylauSF) [inline lib/merch_calendar/fiscal_year_calendar.rb]: Will do! 👍

File: lib/merch_calendar/retail_calendar.rb (hunk 2 above)
Comment (chylauSF) [inline lib/merch_calendar/retail_calendar.rb]: 👍

File: lib/merch_calendar/retail_calendar.rb (hunk 3 above)
Comment (chylauSF) [inline lib/merch_calendar/retail_calendar.rb]: 👍

**Comment by chylauSF** [comment]
I am going to close this PR and re-write a new PR to be more generic, with the changes stated above, as well as add more test coverage.
